### DISCUSSION TITLE
Ability to set a unique short link to each user

### DISCUSSION THREAD

**Discussion by LeaveitT0uchless** [urania-dev/snapp#56] [Ideas]
It would be nice to be able to assign a short link to each user so that there are less -1, -2, etc.

In this case, user1 is assigned to shortlink1 when adding a shortlink regardless of what they put on the short code, it will be /shortlink1[shortcode].

This would make it so each use is assigned a 2+ shortcode.

**Comment by urania-dev** [comment]
To be fair, i did not understand the requests 🧐

**Comment by LeaveitT0uchless** [comment]
Sorry about that 😅. Ok, so here is my use case with the help of chatgpt

Use Case: User-Specific Shortlink Assignment System

Overview:
Each user in the system will be assigned a unique shortlink prefix. Whenever a user creates a new shortlink, the system will automatically prepend their unique prefix to the shortlink code they specify, ensuring no overlap with shortlinks created by other users. This structure avoids duplicates to show up as numbers like -1, -2 to differentiate similar shortlinks.

Actors:

	•	User: An individual who is assigned a unique shortlink prefix.
	•	Admin: Manages the system and assigns unique prefixes to users.
	•	System: The platform handling shortlink creation, storage, and management.

Preconditions:

	1.	Each user is registered and assigned a unique shortlink prefix (e.g., /shortlink1, /shortlink2).
	2.	The system supports user-specific URL generation with a standard structure of /<user_prefix><shortcode>.

Main Flow: Creating a Shortlink

	1.	User Login:
The user logs into their account on the system. Each user is already pre-assigned a unique prefix (e.g., shortlink1 for User1).
	2.	Shortlink Request:
The user initiates a request to create a new shortlink by inputting a custom shortcode they want (e.g., “promo”).
	3.	System Behavior:
	•	Prefix Assignment: The system automatically appends the user’s unique prefix (/shortlink1) to the custom shortcode.
	•	Shortlink Generation: The resulting shortlink would be in the format /shortlink1promo regardless of the shortcode entered.
	4.	Shortlink Validation:
The system ensures that /shortlink1promo does not already exist in the system (to avoid duplicates for that specific user).
	5.	Shortlink Storage:
The system stores the newly created shortlink, associating it with the user’s profile.
	6.	Response:
The system confirms the successful creation of the shortlink /shortlink1/promo.
	7.	Usage:
The user can now share /shortlink1/promo as their custom link. Any visitors who access the link are redirected to the associated destination URL.

Postconditions:

	•	The user has a working shortlink that incorporates their unique prefix.
	•	The system has avoided any shortlink collisions or the need for appending incremental numbers like -1, -2.

**Comment by urania-dev** [comment]
Wouldn't be more aesthetic /username/link at this point?

**Reply by LeaveitT0uchless** [reply]
aesthetically, yes, however, at least for me, the main goal of using this is to have the least possible characters on the link. For 2 reasons, keeping it as short of a link as possible but also making the qr code generated with less squares.

**Reply by urania-dev** [reply]
so it would be a random generated string? (possible for the admins to change or fixed?)
would also it be  subpath or in the url itself? 
```
   snapp.li/prefix-shortcode 
```
or
```
   snapp.li/prefix/shortcode
``` 

i could use a random string generator for that, how long should it be?

**Reply by LeaveitT0uchless** [reply]
I assign a very short 3-6 code to my users, I don't see why it would be needed to limit it through. So user1 has prefix1, then any shortcodes on their end, no matter what will be prefix1[shortcode they added]. Admins can edit this assigned prefix as needed.

it would be nice if they can be assigned multiple and they can pick the prefix to use when creating the shortcode but I can see how this can be complex. Lastly, it would also be cool if multiple users can be assigned to the same prefix and then they can see each other's shortcodes but this goes into more complex user management. For now, I would just have 1 login for everyone that can edit that prefix but then the issue would be that I can then not use 2fa for them.

**Reply by urania-dev** [reply]
At the moment there is already a role based distinction. Usually admin can see everything and user only own content. 

It seems to me that this will be a little ambiguous tho. 

It would work like:
- user can choose from different prefix list assigned to that user? 

- there should be an internal shared state for member of the same prefix? 


Maybe an actual use example would be Better?

I'm registered as user with prefix "develop"|"ui"|"social media"

Those prefix are manually assigned on invitation by admin? Should they be free field or a select to reduce errors and duplicates?

If then i create a ui/ shortcode everyone in the UI/ prefix should see it 

Like colleague of the same office?

**Reply by LeaveitT0uchless** [reply]
Definitely ambitious, for now, it would be nice to just at least get the prefixes implemented, thinking on the future implementation to make that easier down the line. This way it can stop users from going wild with what they create thus limiting others from being able to use those prefixes. These should be free field since that prefix helps tells me what that user manages or where I believe I'm being lead to. 

The idea is that prefixes will be manually assigned by an admin. For example, User 1 might start with the prefixes “develop,” “ui,” and “social media.” Later, they might be granted the “frontend” prefix, allowing them to use it for creating shortcodes. Conversely, an admin could remove a prefix like “frontend,” preventing the user from using or editing it going forward.

I understand this adds complexity, especially when it comes to enabling and disabling prefixes for users. However, keeping this functionality in mind during the initial implementation will make it easier to incorporate in the future.

**Reply by urania-dev** [reply]
well to be honest i'm trying to rationalize it in order to better implement it, and i'm trying to figure out which uses could it hve in order to understand how to integrate it better :)

if it something like a "tag" system that will generate "tagged" shortcode as in ui/shortcode develop/shortcody and so on i think i can implement a management system in which tags get generated, which one are default one, and who can partecipate see them

**Comment by urania-dev** [comment]
so after discussing it this will be released as a setting to be enabled via ui too,
i'm thinking about calling it tags or department,

- the idea would be to have a list of customizable prefix to prepend to urls.
- the admin will choose from a checkbox list which prefix an user should use
- the user will be able to choose among the available one
- so the final generation would be domain/prefix/shortcode
- in case of duplicates prefix/shortcode prefix/shortcode-1 prefix/shortcode-2 prefix/shortcode-3 will follow
- users will be able to browser other shortcode via common prefix. i would leave the browsability to the list only, metrics and notes should be private by design, but still they would be able to see prefix/shortcode and visit where it lands

is that all/correct? :)

**Reply by LeaveitT0uchless** [reply]
Yess this is great! Would it be possible to later make it a toggle between sharing metrics/notes? 

I can see a use case where a user writes a note like: DON'T change this for any reason and another user not seeing it and changing it.

The other use case on the metrics being shared if assigned the same prefix is that some users might care about those more than the others, I can also see users asking to only allow another user to read those metrics and not be able to make changes. Maybe the ability to have usage reports could take care of this though.

**Comment by urania-dev** [comment]
So i tought about this extensively in these days, even tried some code, i think thou that there is compromise i could find, but even if it may not the best solution for you it's the only viable path which i find useful as an update that will target the community and not the single use case.

Have link prepended by a tag chose as above, and that's ok, but it will not remove the uniqueness of the slug, as it would require a rework of the api as a whole for a different scope of the platform itself. so i don't think is a suitable path.

The only compromise as i said would be having a ENV variable like TAGS_AS_PREFIXES (as it is a very radical change in the app flow) 

#### If active: 
- the tags would be assigned in a many-to-one relations (many snapps to one tag only each)
- will stop shortcode visibility without a prefix, 
- the prefixes (tags) would be admin generated and assigned only
- the url will be available at prefix/shortcode
- the conflict and the overwriting of the shortcode with dash-plus-number will be automatic even if they belong to different prefix
- should users belong to a single tags or multiple tags? (es. many user in tag A but only TAG A for said users, or they can choose among different?)
so in case of `prefixA/shortcodeA` and `prefixB/shortcodeA-1` the result would be

| prefix | shortcode | conflict | found at prefix/shortcode |
|-------|-----------|----------|-------------------------|
| prefixA | shortcodeA |  NO | YES |
| prefixB | shortcodeA | YES | NO |
| prefixB | shortcodeA-1| NO | YES |

#### in case the env is inactive 

- the tags would be assigned in a many-to-many relations (many snapps could have many tags)
- the shortcode would be available via both /shortcode /prefix/shortcode
- the conflict and the overwriting of the shortcode with dash-plus-number will be automatic

**Comment by LeaveitT0uchless** [comment]
Thank you for looking into this!

Yes, I believe it’s better to make sure this can be used by the community. The approach seems unique compared to anything I’ve come across, and I think the community will find many use cases for it.

In terms of user management, I believe **users should belong to multiple tags**. Based on your implementation, this will allow for more flexibility. Since I want to keep the links as short as possible, the way I’d use this would be: the tags would consist of 3 letters—each specifying the **company’s first letter**, **location’s first letter**, and **department’s first letter**. From there, I’d only need to ensure the shortcode itself is unique.

For example, if the company is *5sushi*, located in New York, and it’s for the HR department, a shortcode for each table could look like: https://snapp.li/5nh/5nht1. This keeps the structure organized and compact.

Looking back, my original request was actually two different requests:

​	1.	**Allow users to edit multiple snapps**, even if they didn’t create them.

​	2.	**A better way to manage shortcodenames**.

I appreciate you addressing the first request. I understand why shortcodes need to remain unique to as they are essentially a unique id. While it would have been great to follow a cleaner structure with shorter links (e.g., https://snapp.li/5nh/t1), I’m okay with focusing on improving user management for now.

**Reply by urania-dev** [reply]
Mh, no wait in my example before edit on snapps remains solely to owner as it Is now. It would simply offer the chance ti navigate tags and there find the other users shortcode within the UI (which otherwise would be viewable by author and admins).

Edit let me rephrase

Atm the owner of urls can see own snapps

Tags would offer the list of other users snapps.in within the app

**Reply by LeaveitT0uchless** [reply]
ah ok, I see, that's still good it'll be a way to introduce the tags and start the separation of users via them.

**Reply by urania-dev** [reply]
I may allow edit from tags page if URL not private (would require a new field with default true) to allow edit through said page but what would be changed by others? The original URL? The shortcode? The notes? That would put user in position to not find snapps previously created if any other user decide to steal shortcode changin original URL or change the snapps alltogether. Would be less ambiguous keep the edit to owner only and admin for moderation/security

**Reply by LeaveitT0uchless** [reply]
I was thinking that the tag the user belongs would essentially allow them to see/edit the shortcodes under that tag so if I belonged to 5nh and 5mh then I can see those shortcodes and add new short codes under them but if I don't belong to any of them, instead I belong to 5th that was just created for me and have no shortcodes my list would be blank and any short codes I make are domain.com/5th/shortcode. 


If that's still alot then we can simply just do it where each user is just assigned to one tag so any tags that user creates will always have that tag (es. domain.com/5nh/shortcode). This essentially makes it so user 5nh cannot create shortcode links for 5mh which is more of a limiting the user kind of feature.

**Reply by urania-dev** [reply]
Tonight / tomorrow i should be able to upload mfa and tags system if that doesnt work i think I can work to department hierarchy decoupled by personal shortcode as you explained in last post 🤔

Apis will require some more time tho

**Comment by urania-dev** [comment]
v.0.8.7 integrates tag, internal second factor for authentication and integration with oidc/keycloak :)
